src/graph.py

```diff
@@ -752,6 +750,104 @@ def perform_GO_enrichment_on_clusters(nxgraph, clustering):
     """
     pass
 
+
+###############################################################################
+            #######            Results           #######
+###############################################################################
+
+def summarize_grid_search(results, mode, top_n=-1): 
+    """
+    Summarizes results of `_grid_randomization` into a matrix where each row is a gene 
+    and each column is a parameter run. If summarizing "membership", entries will be 0 or 1
+    indicating whether or not a node appeared in each experiment. If summarizing "robustness"
+    or "specificity", entries indicate robustness or specificity values for each experiment. 
+    Also, node attributes columns are added for plotting purposes.
+
+    Arguments:
+        results (list of tuples): Results of `_grid_randomization` (paramstring, forest, augmented forest)
+        mode (str): Reported values "membership", "robustness", "specificity"
+        top_n (int): Takes the top_n values of the summary dataframe. top_n=-1 sets no threshold
+
+    Returns:
+        pd.DataFrame: Columns correspond to each parameter experiment, indexed by nodes
+    """
+    
+    # Exclude any degenerate results
+    results = {paramstring: graphs for paramstring, graphs in results.items() if graphs["augmented_forest"].number_of_nodes() > 0}
+    
+    if mode == "membership": # Summarize single-run parameter search
+        dfs = [pd.Series(1, index=graphs["augmented_forest"].nodes(), name=paramstring) for paramstring, graphs in results.items()]
+    elif mode == "robustness": # Summarize randomized robustness
+        dfs = [get_networkx_graph_as_dataframe_of_nodes(graphs["augmented_forest"])["robustness"].rename(paramstring) for paramstring, graphs in results.items()]
```

Comment:
isn't this a list of series you're naming "dfs"?

Reply:
aba1f17644e68cd23faae2706dec8e02854f89a1